src/sentry/sentry_metrics/indexer/postgres_v2.py

```diff
@@ -0,0 +1,277 @@
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+)
+
+from django.db.models import Q
+
+from sentry.sentry_metrics.indexer.cache import indexer_cache
+from sentry.sentry_metrics.indexer.models import StringIndexer as StringIndexerTable
+from sentry.utils import metrics
+from sentry.utils.services import Service
+
+_INDEXER_CACHE_METRIC = "sentry_metrics.indexer.memcache"
+
+
+class KeyCollection:
+    """
+    A KeyCollection is a way of keeping track of a group of keys
+    used to fetch ids, which are stored as results. There are
+    three potential steps to fetching the ids, and each step
+    can utilize a new KeyCollection:
+        1. ids from cache
+        2. ids from existing db records
+        3. ids from newly created db records
+
+    A key is a org_id, string pair, either represented as a
+    tuple e.g (1, "a"), or a string "1:a".
+
+    Initial mapping is org_id's to sets of strings:
+        { 1: {"a", "b", "c"}, 2: {"e", "f"} }
+
+    Resulting ids will be stored in self.results, which is a
+    mapping of org_id to string -> int mapping:
+        {
+            1: {"a": 10, "b": 11, "c": 12},
+            2: {"e": 13},
+        }
+
+    For any keys that we didn't find ids for in the current
+    step (and therefore weren't added to the results) , use
+    `get_unmapped_keys` to return a new KeyCollection that
+    can be used for the next step in getting those ids.
+    """
+
+    def __init__(self, mapping: Mapping[int, Set[str]]):
+        self.mapping = mapping
+
+    def as_tuples(self) -> MutableSequence[Tuple[int, str]]:
+        """
+        Returns all the keys, each key represented as tuple -> (1, "a")
+        """
+        key_pairs: MutableSequence[Tuple[int, str]] = []
+        for org_id in self.mapping:
+            key_pairs.extend([(org_id, string) for string in self.mapping[org_id]])
+
+        return key_pairs
+
+    def as_strings(self) -> MutableSequence[str]:
+        """
+        Returns all the keys, each key represented as string -> "1:a"
+        """
+        keys: MutableSequence[str] = []
+        for org_id in self.mapping:
+            keys.extend([f"{org_id}:{string}" for string in self.mapping[org_id]])
+
+        return keys
+
+
+KR = TypeVar("KR", bound="KeyResult")
+
+
+@dataclass(frozen=True)
+class KeyResult:
+    org_id: int
+    string: str
+    id: int
+
+    @classmethod
+    def from_string(cls: Type[KR], key: str, id: int) -> KR:
+        org_id, string = key.split(":")
+        return cls(int(org_id), string, id)
+
+
+class KeyResults:
+    def __init__(self) -> None:
+        self.results: MutableMapping[int, MutableMapping[str, int]] = {}
+
+    def add_key_results(self, key_results: Sequence[KeyResult]) -> None:
+        for key_result in key_results:
+            try:
+                self.results[key_result.org_id][key_result.string] = key_result.id
+            except KeyError:
+                self.results[key_result.org_id] = {key_result.string: key_result.id}
```

Comment:
Please, either use a defaultdict so you do not need to initialize the missing keys or check the presence of your key before trying to store. Catching the exception is unnecessarily more expensive.